### PR TITLE
Always use `importlib_metadata.version` when obtaining package version to prevent version mismatch between `importlib_metadata.version` and `<package>.__version__`

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## What changes are proposed in this pull request?

Always use `importlib_metadata.version` when obtaining a package version to prevent version mismatch between `importlib_metadata.version` (used in auto dependency inference) and `<package>.__version__` (used in `get_default_pip_requirements`).

## How is this patch tested?

Existing tests

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/requirements_utils.py [hunk 1]
```diff
@@ -180,17 +180,30 @@ def _run_command(cmd):
...
-
     # In Databricks, strip a dev version suffix for pyspark (e.g. '3.1.2.dev0' -> '3.1.2')
     # and make it installable from PyPI.
     if package == "pyspark" and is_in_databricks_runtime():
-        version = _strip_dev_version_suffix(version)
-    return version
+        return _strip_dev_version_suffix(__import__("pyspark").__version__)
+
+    try:
+        return importlib_metadata.version(package)
```
Comment (harupy) [inline mlflow/utils/requirements_utils.py:194]: If a user installs `torch` with:

```
pip install torch==1.8.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
```

This line returns `1.8.0`.

File: mlflow/utils/requirements_utils.py [hunk 2]
```diff
@@ -180,16 +180,31 @@ def _run_command(cmd):
...
-def _get_package_version_from_metadata(package):
+def _get_installed_version(package, module=None):
     """
-    Obtains the version of the specified package from its metadata.
+    Obtains the package version using `importlib_metadata.version`. If it fails, use
+    `__import__(module or package).__version__`.
     """
-    version = importlib_metadata.version(package)
+    try:
+        version = importlib_metadata.version(package)
```
Comment (harupy) [inline mlflow/utils/requirements_utils.py:189]: If a user installs `torch` with:

```
pip install torch==1.8.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
```

`importlib_metadata.version("torch")` returns `1.8.0`.

Comment (harupy) [reply mlflow/utils/requirements_utils.py:189]: Or we could just log `torch==1.8.0+cpu`. If the user wants to specify `--find-links`,  he/she can use `pip_requirements`.

File: mlflow/utils/requirements_utils.py [hunk 3]
```diff
@@ -180,16 +180,31 @@ def _run_command(cmd):
...
+        # `__import__(package).__version__`. See the example for pytorch below.
+        #
+        # Example
+        # -------
+        # $ pip install torch==1.9.0
+        # $ python -c "import torch; print(torch.__version__)"
+        # 1.9.0+cu102
+        # $ python -c "import importlib_metadata; print(importlib_metadata.version('torch'))"
+        # 1.9.0
+        version = __import__(module or package).__version__
```
Comment (harupy) [inline mlflow/utils/requirements_utils.py:201]: The best effort to capture a package version.

**Review by dbczumar — COMMENTED** [review]
@harupy Do you have an example of the mismatch? Would help to understand its significance.

**Comment by harupy** [comment]
@dbczumar Sure!

```bash
% python -c "import mxnet; print(mxnet.__version__)"
1.8.0
                                                                                                                                                                        
% python -c "import importlib_metadata; print(importlib_metadata.version('mxnet'))"
1.8.0.post0
```

I think the same thing happens for `torch`.

File: mlflow/utils/requirements_utils.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/utils/requirements_utils.py:189]: @harupy Ideally, we'd like to maximize the likelihood of serving succeeding. Can we ignore local version identifiers and print a warning indicating that the local version identifier was dropped & instructing customers to use `pip_requirements` or `conda_env` to specify the distribution containing the local version if necessary?

Comment (harupy) [reply mlflow/utils/requirements_utils.py:189]: Ok, I'll update the code to drop local version identifiers and raise a warning for that.

File: mlflow/utils/requirements_utils.py [hunk 4]
```diff
@@ -240,12 +258,24 @@ def _infer_requirements(model_uri, flavor):
...
+    :param version: A version string.
+    """
+    try:
+        return Version(version).local
+    except InvalidVersion:
+        return None
 
 
-def _strip_local_version_identifier(version):
+def _strip_local_version_label(version):
```
Comment (harupy) [inline mlflow/utils/requirements_utils.py:276]: Renamed `_strip_local_version_identifier` to `_strip_local_version_label`. The following code describes why:

```
1.2.3+abc
     ^^^^ local version label (segment starting with '+')
^^^^^^^^^ local version identifier (version identifier containing a local version label)
```

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @harupy !

**Pull request merged**